File: telemetry/aitf/sdk/typescript/src/exporters/immutable-log.ts
/**
 * AITF Immutable Log Exporter.
 *
 * OTel SpanExporter that writes AI telemetry events to an append-only,
 * hash-chained log file providing cryptographic tamper evidence.
 *
 * Each log entry includes a SHA-256 hash of the previous entry, creating
 * an unbroken chain. Any modification to a historical entry invalidates
 * all subsequent hashes, making tampering detectable.
 *
 * Satisfies audit requirements for:
 *   - EU AI Act Article 12 (record-keeping)
 *   - NIST AI RMF GOVERN-1.5 (audit trail)
 *   - SOC 2 CC8.1 (integrity)
 *   - ISO/IEC 42001 (AI management records)
 *
 * Usage:
 *   import { ImmutableLogExporter } from "@aitf/sdk";
 *   const exporter = new ImmutableLogExporter({
 *     logFile: "/var/log/aitf/immutable_audit.jsonl",
 *   });
 */

import * as crypto from "crypto";
import * as fs from "fs";
import * as path from "path";
import { ReadableSpan, SpanExporter } from "@opentelemetry/sdk-trace-base";
import { ExportResult, ExportResultCode } from "@opentelemetry/core";
import { OCSFMapper } from "../ocsf/mapper";
import { ComplianceMapper } from "../ocsf/compliance-mapper";
import { AIBaseEvent, stripNulls } from "../ocsf/schema";

const GENESIS_HASH =
  "0000000000000000000000000000000000000000000000000000000000000000";
const MAX_FILE_SIZE = 1024 * 1024 * 1024; // 1 GB

/** Compute SHA-256 hash for a log entry. */
function computeEntryHash(
  seq: number,
  timestamp: string,
  prevHash: string,
  eventJSON: string
): string {
  const payload = `${seq}|${timestamp}|${prevHash}|${eventJSON}`;
  return crypto.createHash("sha256").update(payload, "utf-8").digest("hex");
}

/** A single entry in the hash-chained log. */
interface ImmutableLogEntry {
  seq: number;
  timestamp: string;
  prev_hash: string;
  hash: string;
  event: Record<string, unknown>;
}

/** Options for ImmutableLogExporter. */
export interface ImmutableLogExporterOptions {
  logFile?: string;
  complianceFrameworks?: string[];
  rotateOnSize?: boolean;
  filePermissions?: number;
}

/**
 * Writes AITF events to an append-only, hash-chained log file.
 */
export class ImmutableLogExporter implements SpanExporter {
  private readonly _logFile: string;
  private readonly _rotateOnSize: boolean;
  private readonly _filePermissions: number;
  private readonly _mapper: OCSFMapper;
  private readonly _complianceMapper: ComplianceMapper;
  private _prevHash: string;
  private _seq: number;
  private _eventCount: number;

  constructor(options: ImmutableLogExporterOptions = {}) {
    this._logFile = path.resolve(
      options.logFile ?? "/var/log/aitf/immutable_audit.jsonl"
    );
    this._rotateOnSize = options.rotateOnSize ?? true;
    this._filePermissions = options.filePermissions ?? 0o600;
    this._mapper = new OCSFMapper();
    this._complianceMapper = new ComplianceMapper({
      frameworks: options.complianceFrameworks,
    });
    this._prevHash = GENESIS_HASH;
    this._seq = 0;
    this._eventCount = 0;

    // Create directory
    const dir = path.dirname(this._logFile);
    fs.mkdirSync(dir, { recursive: true });

    // Resume chain
    this._resumeChain();
  }

  private _resumeChain(): void {
    if (!fs.existsSync(this._logFile)) return;

    try {
      const content = fs.readFileSync(this._logFile, "utf-8");
      const lines = content.trim().split("\n").filter(Boolean);
      if (lines.length === 0) return;

      const lastLine = lines[lines.length - 1];
      const entry: ImmutableLogEntry = JSON.parse(lastLine);
      this._prevHash = entry.hash;
      this._seq = entry.seq + 1;
      console.log(
        `immutable_log: resumed chain at seq=${this._seq} from ${this._logFile}`
      );
    } catch {
      this._prevHash = GENESIS_HASH;
      this._seq = 0;
    }
  }

  export(
    spans: ReadableSpan[],
    resultCallback: (result: ExportResult) => void
  ): void {
    const entries: string[] = [];

    for (const span of spans) {
      const ocsfEvent = this._mapper.mapSpan(span);
      if (!ocsfEvent) continue;

      // Enrich with compliance metadata
      const eventType = this._classifyEvent(ocsfEvent);
      if (eventType) {
        this._complianceMapper.enrichEvent(ocsfEvent, eventType);
      }

      if (
        typeof ocsfEvent !== "object" ||
        ocsfEvent === null ||
        Array.isArray(ocsfEvent)
      ) {
        continue;
      }
      const eventDict = stripNulls(
        ocsfEvent as unknown as Record<string, unknown>
      );

      const eventJSON = JSON.stringify(eventDict, Object.keys(eventDict).sort());
      const timestamp = new Date().toISOString();
      const entryHash = computeEntryHash(
        this._seq,
        timestamp,
        this._prevHash,
        eventJSON
      );

      const entry: ImmutableLogEntry = {
        seq: this._seq,
        timestamp,
        prev_hash: this._prevHash,
        hash: entryHash,
        event: eventDict,
      };

      entries.push(JSON.stringify(entry));
      this._prevHash = entryHash;
      this._seq++;
      this._eventCount++;
    }

    if (entries.length === 0) {
      resultCallback({ code: ExportResultCode.SUCCESS });
      return;
    }

    try {
      this._writeEntries(entries);
      resultCallback({ code: ExportResultCode.SUCCESS });
    } catch (err) {
      console.error("immutable_log: write failed:", err);
      resultCallback({ code: ExportResultCode.FAILED });
    }
  }

  private _writeEntries(entries: string[]): void {
    // Check rotation
    if (this._rotateOnSize && fs.existsSync(this._logFile)) {
      const stats = fs.statSync(this._logFile);
      if (stats.size > MAX_FILE_SIZE) {
        const rotated =
          this._logFile +
          "." +
          new Date().toISOString().replace(/[:.]/g, "");
        fs.renameSync(this._logFile, rotated);
      }
    }

    // Append with restrictive permissions
    const content = entries.join("\n") + "\n";
    fs.appendFileSync(this._logFile, content, {
      encoding: "utf-8",
      mode: this._filePermissions,
    });
  }

  /**
   * Classify an OCSF event for compliance mapping.
   *
   * AITF reuses existing OCSF classes (OCSF PR #1641 / issue #1640), so
   * `class_uid` is no longer unique per AI event type (e.g. inference and
   * tool execution both reuse API Activity 6003). Classify by the
   * discriminating fields the mapper sets on each event.
   */
  private _classifyEvent(event: AIBaseEvent): string | null {
    const e = event as unknown as Record<string, unknown>;
    if ("model" in e) return "model_inference";
    if ("tool_name" in e) return "tool_execution";
    if ("database_name" in e) return "data_retrieval";
    if ("finding" in e) return "security_finding";
    if ("session_id" in e && "agent_id" in e) return "agent_activity";
    return null;
  }

  get eventCount(): number {
    return this._eventCount;
  }

  get currentSeq(): number {
    return this._seq;
  }

  get currentHash(): string {
    return this._prevHash;
  }

  async shutdown(): Promise<void> {
    // No-op
  }

  async forceFlush(): Promise<void> {
    // No-op
  }
}

/** Result of an immutable log verification. */
export interface VerificationResult {
  valid: boolean;
  entriesChecked: number;
  firstInvalidSeq?: number;
  expectedHash?: string;
  foundHash?: string;
  finalHash?: string;
  error?: string;
}

/**
 * Verify the integrity of an AITF immutable log file
 * by replaying the entire hash chain from genesis.
 */
export function verifyImmutableLog(logFile: string): VerificationResult {
  if (!fs.existsSync(logFile)) {
    return {
      valid: false,
      entriesChecked: 0,
      error: "Log file does not exist",
    };
  }

  const content = fs.readFileSync(logFile, "utf-8");
  const lines = content.trim().split("\n").filter(Boolean);

  let prevHash = GENESIS_HASH;
  let entriesChecked = 0;

  for (const line of lines) {
    let entry: ImmutableLogEntry;
    try {
      entry = JSON.parse(line);
    } catch (err) {
      return {
        valid: false,
        entriesChecked,
        firstInvalidSeq: entriesChecked,
        error: `Invalid JSON at seq ${entriesChecked}: ${err}`,
      };
    }

    // Check chain linkage
    if (entry.prev_hash !== prevHash) {
      return {
        valid: false,
        entriesChecked,
        firstInvalidSeq: entry.seq,
        expectedHash: prevHash,
        foundHash: entry.prev_hash,
        error: `Chain break at seq ${entry.seq}: prev_hash mismatch`,
      };
    }

    // Recompute hash
    const eventJSON = JSON.stringify(
      entry.event,
      Object.keys(entry.event).sort()
    );
    const computedHash = computeEntryHash(
      entry.seq,
      entry.timestamp,
      entry.prev_hash,
      eventJSON
    );

    if (computedHash !== entry.hash) {
      return {
        valid: false,
        entriesChecked,
        firstInvalidSeq: entry.seq,
        expectedHash: computedHash,
        foundHash: entry.hash,
        error: `Hash mismatch at seq ${entry.seq}: entry tampered`,
      };
    }

    prevHash = entry.hash;
    entriesChecked++;
  }

  return {
    valid: true,
    entriesChecked,
    finalHash: prevHash,
  };
}

File: telemetry/aitf/sdk/go/exporters/immutable_log.go
// Package exporters provides AITF span exporters for Go.
//
// ImmutableLogExporter writes AI telemetry events to an append-only,
// hash-chained log file providing cryptographic tamper evidence.
//
// Each log entry includes a SHA-256 hash of the previous entry, creating
// an unbroken chain. Any modification to a historical entry invalidates
// all subsequent hashes, making tampering detectable.
//
// Satisfies audit requirements for:
//   - EU AI Act Article 12 (record-keeping)
//   - NIST AI RMF GOVERN-1.5 (audit trail)
//   - SOC 2 CC8.1 (integrity)
//   - ISO/IEC 42001 (AI management records)
package exporters

import (
	"bufio"
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/girdav01/AITF/sdk/go/ocsf"

	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

const (
	genesisHash    = "0000000000000000000000000000000000000000000000000000000000000000"
	maxLogFileSize = 1024 * 1024 * 1024 // 1 GB
)

// computeEntryHash computes the SHA-256 hash for a log entry.
func computeEntryHash(seq int, timestamp, prevHash, eventJSON string) string {
	payload := fmt.Sprintf("%d|%s|%s|%s", seq, timestamp, prevHash, eventJSON)
	h := sha256.Sum256([]byte(payload))
	return fmt.Sprintf("%x", h)
}

// ImmutableLogEntry represents a single entry in the hash-chained log.
type ImmutableLogEntry struct {
	Seq       int                    `json:"seq"`
	Timestamp string                 `json:"timestamp"`
	PrevHash  string                 `json:"prev_hash"`
	Hash      string                 `json:"hash"`
	Event     map[string]interface{} `json:"event"`
}

// ImmutableLogExporter writes hash-chained log entries for tamper evidence.
type ImmutableLogExporter struct {
	logFile    string
	mapper     *ocsf.OCSFMapper
	compliance *ocsf.ComplianceMapper
	prevHash   string
	seq        int
	eventCount int
	mu         sync.Mutex
}

// ImmutableLogOption configures the ImmutableLogExporter.
type ImmutableLogOption func(*ImmutableLogExporter)

// WithImmutableLogFile sets the log file path.
func WithImmutableLogFile(path string) ImmutableLogOption {
	return func(e *ImmutableLogExporter) { e.logFile = path }
}

// WithImmutableComplianceFrameworks sets compliance frameworks for enrichment.
func WithImmutableComplianceFrameworks(frameworks []string) ImmutableLogOption {
	return func(e *ImmutableLogExporter) {
		e.compliance = ocsf.NewComplianceMapper(frameworks)
	}
}

// NewImmutableLogExporter creates a new immutable log exporter.
func NewImmutableLogExporter(opts ...ImmutableLogOption) (*ImmutableLogExporter, error) {
	e := &ImmutableLogExporter{
		logFile:    "/var/log/aitf/immutable_audit.jsonl",
		mapper:     ocsf.NewOCSFMapper(),
		compliance: ocsf.NewComplianceMapper(nil),
		prevHash:   genesisHash,
		seq:        0,
	}
	for _, opt := range opts {
		opt(e)
	}

	// Create directory
	dir := filepath.Dir(e.logFile)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create log directory %s: %w", dir, err)
	}

	// Resume chain from existing log
	e.resumeChain()

	return e, nil
}

func (e *ImmutableLogExporter) resumeChain() {
	f, err := os.Open(e.logFile)
	if err != nil {
		return // File doesn't exist yet
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 10*1024*1024) // 10MB max line

	var lastLine string
	lineCount := 0
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			lastLine = line
			lineCount++
		}
	}

	if lastLine != "" {
		var entry ImmutableLogEntry
		if err := json.Unmarshal([]byte(lastLine), &entry); err == nil {
			e.prevHash = entry.Hash
			e.seq = entry.Seq + 1
			log.Printf("immutable_log: resumed chain at seq=%d from %s", e.seq, e.logFile)
		}
	}
}

// ExportSpans converts spans to hash-chained log entries.
func (e *ImmutableLogExporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpan) error {
	var entries []string

	for _, span := range spans {
		ocsfEvent := e.mapper.MapSpan(span)
		if ocsfEvent == nil {
			continue
		}

		eventBytes, err := json.Marshal(ocsfEvent)
		if err != nil {
			continue
		}

		var eventMap map[string]interface{}
		if err := json.Unmarshal(eventBytes, &eventMap); err != nil {
			continue
		}

		// Canonical JSON for deterministic hashing
		canonicalBytes, err := json.Marshal(eventMap)
		if err != nil {
			continue
		}
		eventJSON := string(canonicalBytes)

		e.mu.Lock()
		timestamp := time.Now().UTC().Format(time.RFC3339Nano)
		entryHash := computeEntryHash(e.seq, timestamp, e.prevHash, eventJSON)

		entry := ImmutableLogEntry{
			Seq:       e.seq,
			Timestamp: timestamp,
			PrevHash:  e.prevHash,
			Hash:      entryHash,
			Event:     eventMap,
		}

		entryBytes, err := json.Marshal(entry)
		if err != nil {
			e.mu.Unlock()
			continue
		}

		entries = append(entries, string(entryBytes))
		e.prevHash = entryHash
		e.seq++
		e.eventCount++
		e.mu.Unlock()
	}

	if len(entries) == 0 {
		return nil
	}

	return e.writeEntries(entries)
}

func (e *ImmutableLogExporter) writeEntries(entries []string) error {
	// Check rotation
	if info, err := os.Stat(e.logFile); err == nil && info.Size() > maxLogFileSize {
		rotated := e.logFile + "." + time.Now().UTC().Format("20060102T150405")
		if err := os.Rename(e.logFile, rotated); err == nil {
			log.Printf("immutable_log: rotated to %s", rotated)
		}
	}

	// Open in append-only mode with restrictive permissions
	f, err := os.OpenFile(e.logFile, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0600)
	if err != nil {
		return fmt.Errorf("immutable_log: open failed: %w", err)
	}
	defer f.Close()

	for _, entry := range entries {
		if _, err := f.WriteString(entry + "\n"); err != nil {
			return fmt.Errorf("immutable_log: write failed: %w", err)
		}
	}

	return f.Sync()
}

// Shutdown is a no-op.
func (e *ImmutableLogExporter) Shutdown(ctx context.Context) error {
	return nil
}

// EventCount returns the total number of events written.
func (e *ImmutableLogExporter) EventCount() int {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.eventCount
}

// CurrentSeq returns the current sequence number.
func (e *ImmutableLogExporter) CurrentSeq() int {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.seq
}

// CurrentHash returns the hash of the last written entry.
func (e *ImmutableLogExporter) CurrentHash() string {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.prevHash
}

// VerificationResult holds the result of a log integrity check.
type VerificationResult struct {
	Valid           bool   `json:"valid"`
	EntriesChecked  int    `json:"entries_checked"`
	FirstInvalidSeq *int   `json:"first_invalid_seq,omitempty"`
	ExpectedHash    string `json:"expected_hash,omitempty"`
	FoundHash       string `json:"found_hash,omitempty"`
	FinalHash       string `json:"final_hash,omitempty"`
	Error           string `json:"error,omitempty"`
}

// VerifyImmutableLog verifies the integrity of an immutable log file
// by replaying the entire hash chain from genesis.
func VerifyImmutableLog(logFile string) (*VerificationResult, error) {
	f, err := os.Open(logFile)
	if err != nil {
		return &VerificationResult{Valid: false, Error: fmt.Sprintf("cannot open file: %v", err)}, nil
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 10*1024*1024)

	prevHash := genesisHash
	entriesChecked := 0

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		var entry ImmutableLogEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			seq := entriesChecked
			return &VerificationResult{
				Valid:           false,
				EntriesChecked:  entriesChecked,
				FirstInvalidSeq: &seq,
				Error:           fmt.Sprintf("invalid JSON at seq %d: %v", entriesChecked, err),
			}, nil
		}

		// Check chain linkage
		if entry.PrevHash != prevHash {
			return &VerificationResult{
				Valid:           false,
				EntriesChecked:  entriesChecked,
				FirstInvalidSeq: &entry.Seq,
				ExpectedHash:    prevHash,
				FoundHash:       entry.PrevHash,
				Error:           fmt.Sprintf("chain break at seq %d: prev_hash mismatch", entry.Seq),
			}, nil
		}

		// Recompute hash
		eventJSON, _ := json.Marshal(entry.Event)
		computedHash := computeEntryHash(entry.Seq, entry.Timestamp, entry.PrevHash, string(eventJSON))

		if computedHash != entry.Hash {
			return &VerificationResult{
				Valid:           false,
				EntriesChecked:  entriesChecked,
				FirstInvalidSeq: &entry.Seq,
				ExpectedHash:    computedHash,
				FoundHash:       entry.Hash,
				Error:           fmt.Sprintf("hash mismatch at seq %d: entry tampered", entry.Seq),
			}, nil
		}

		prevHash = entry.Hash
		entriesChecked++
	}

	return &VerificationResult{
		Valid:          true,
		EntriesChecked: entriesChecked,
		FinalHash:     prevHash,
	}, nil
}

// Compile-time check that ImmutableLogExporter implements SpanExporter.
var _ sdktrace.SpanExporter = (*ImmutableLogExporter)(nil)
